### PR TITLE
Upgrade version

### PR DESCRIPTION
**Pull request by adilhussain540**
Upgrade to the latest available version.

### CODE REVIEW

**Review by lrozema — APPROVED** [review]
Hard to tell if this is good or bad, I guess we can always roll back if this breaks things.

**Comment by lrozema** [comment]
Once this gets merged we have to follow these steps to get it into production: https://www.notion.so/omnidots/AWS-server-management-054806150d6742ca9e8502b62dfaad49?pvs=4#4066634194894f96835e0cdebbe4a53c

File: exactonline/api/autorefresh.py [hunk 1]
```diff
@@ -1,26 +1,66 @@
...
+    > Use your refresh token to get a new access token before it
+    > expires. If you make an API call with an expired access token, your
+    > call will be rejected with response code 401 and reason
+    >
+    > **Do not request a new access token too early**
+    > You should only request a new token right before it expires. You
+    > can only request a new access token 9 minutes and 30 seconds after
+    > you received it. If you request an access token before the 9 minutes
+    > and 30 seconds have passed, your call will be rejected with response
+    > code 401 and reason Access Token not expired.
```
Comment (AhsanPioneer) [inline exactonline/api/autorefresh.py:31]:
```suggestion
    > code 401 and reason 'Access Token not expired'.
```

File: exactonline/api/autorefresh.py [hunk 2]
```diff
@@ -1,26 +1,66 @@
...
         try:
             decoded = super(Autorefresh, self).rest(request)
         except HTTPError as e:
-            if e.code != 401:
-                raise
+            if e.code == 401 and not have_fresh_token:
+                # If we received a 401 even though we think our token is
+                # still valid, maybe we were wrong about the expiry
+                # time. (Maybe one of the clocks is off, maybe the
+                # remote side flushed the tokens.)
```
Comment (AhsanPioneer) [inline exactonline/api/autorefresh.py:57]: Wouldn't it be better to mention these reasons as bullets

File: exactonline/api/bulksalesitemprices.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+class BulkSalesItemPrices(Manager):
+    resource = 'bulk/Logistics/SalesItemPrices'
+    # https://start.exactonline.co.uk/docs/HlpRestAPIResourcesDetails.aspx?
+    #   name=BulkLogisticsSalesItemPrices
+
+    def filter(self, item_id=None, **kwargs):
+        if 'select' not in kwargs:
+            kwargs['select'] = (
+                'ID,Account,AccountName,Item,ItemCode,Price,Currency')
```
Comment (AhsanPioneer) [inline exactonline/api/bulksalesitemprices.py:20]:
```suggestion
        kwargs['select'] = kwargs.get('select', 'ID,Account,AccountName,Item,ItemCode,Price,Currency')
```

File: exactonline/api/items.py [hunk 4]
```diff
@@ -1,15 +1,32 @@
...
 from .manager import Manager
 
 
 class Items(Manager):
+    # https://start.exactonline.co.uk/docs/HlpRestAPIResourcesDetails.aspx?
+    #   name=LogisticsItems
     resource = 'logistics/Items'
 
-    def filter(self, code=None, **kwargs):
+    def filter(self, code=None, modified_since_date=None, **kwargs):
```
Comment (AhsanPioneer) [inline exactonline/api/items.py:17]: Thinking out loud
For each filter function we allow using optional arguments which would serve as filters
How about we modify the base filter function itself to accept a `filter_dict`
In that way we would be able to send params like 
`{"code": u'Code eq  %s' % (code,)}`



And then can do this in the Manager `filter` method
```
for filter_key, filter_value in filter_dict:
     self._filter_append(kwargs, filter_value)
```

File: exactonline/http.py [hunk 5]
```diff
@@ -232,36 +234,22 @@ class CustomValidHTTPSConnection(ValidHTTPSConnection):
...
     """
-    return _http_request(url, method='POST', data=_marshalled(data), opt=opt)
-
+    if method in ('DELETE', 'GET'):
+        assert data is None, (method, url, data)
+    elif method in ('POST', 'PUT'):
+        pass
+    else:
+        raise NotImplementedError(
+            'No REST handler for method %s' % (method,))
```
Comment (AhsanPioneer) [inline exactonline/http.py:249]:
```suggestion

    raise NotImplementedError('No REST handler for method %s' % (method,))
```

File: exactonline/http.py [hunk 6]
```diff
@@ -276,7 +264,29 @@ def _marshalled(data):
...
+                # X-RateLimit-Remaining: 8924
+                remaining=headers.get('x-ratelimit-remaining'))
+        if headers.get('x-ratelimit-minutely-reset'):
+            limiter.update(
+                # X-RateLimit-Minutely-Reset: 1638447360000
+                until=headers.get('x-ratelimit-minutely-reset'),
+                # X-RateLimit-Minutely-Limit: 100
+                limit=headers.get('x-ratelimit-minutely-limit'),
+                # X-RateLimit-Minutely-Remaining: 99
+                remaining=headers.get('x-ratelimit-minutely-remaining'))
```
Comment (AhsanPioneer) [inline exactonline/http.py:286]:
```suggestion
    if not limiter:
        return

    # First Daily, then Minutely. If both keys are the same, the shortest
    # should win. And the minute timer is likely the shortest.
    if headers.get('x-ratelimit-reset'):
        limiter.update(
            # X-RateLimit-Reset: 1638489600000
            until=headers.get('x-ratelimit-reset'),
            # X-RateLimit-Limit: 9000
            limit=headers.get('x-ratelimit-limit'),
            # X-RateLimit-Remaining: 8924
            remaining=headers.get('x-ratelimit-remaining'))
    if headers.get('x-ratelimit-minutely-reset'):
        limiter.update(
            # X-RateLimit-Minutely-Reset: 1638447360000
            until=headers.get('x-ratelimit-minutely-reset'),
            # X-RateLimit-Minutely-Limit: 100
            limit=headers.get('x-ratelimit-minutely-limit'),
            # X-RateLimit-Minutely-Remaining: 99
            remaining=headers.get('x-ratelimit-minutely-remaining'))
```

can avoid the indentation

File: exactonline/http.py [hunk 7]
```diff
@@ -276,7 +264,29 @@ def _marshalled(data):
     return data
 
 
-def _http_request(url, method=None, data=None, opt=None):
+def _update_ratelimiter_with_exactonline_headers(limiter, headers):
+    if limiter:
+        # First Daily, then Minutely. If both keys are the same, the shortest
+        # should win. And the minute timer is likely the shortest.
+        if headers.get('x-ratelimit-reset'):
```
Comment (AhsanPioneer) [inline exactonline/http.py:271]: can use constants for these header keys which appear more than once

File: exactonline/http.py [hunk 8]
```diff
@@ -308,6 +321,14 @@ def _http_request(url, method=None, data=None, opt=None):
         stored_exception = exception
     finally:
         if fp:
+            # Store ratelimit values if available.
+            if limiter:
```
Comment (AhsanPioneer) [inline exactonline/http.py:325]: Don't need an if check here we already check for it in the function

File: exactonline/http_test.py [hunk 9]
```diff
@@ -51,47 +58,65 @@ def __init__(self, method, code, body=None, use_ssl=False):
...
         try:
             peersock, peeraddr = self.socket.accept()
         except ssl.SSLError:
             # Broken connection by peer.
             return
 
         data = peersock.recv(4096)
-        if HttpTestCase.to_str(data).startswith(self.method + ' '):
-            if self.body is None:
+        if HttpTestCase.to_str(data).startswith(response.method + ' '):
```
Comment (AhsanPioneer) [inline exactonline/http_test.py:83]:
```suggestion
        if HttpTestCase.to_str(data).startswith(f"response.method "):
```

File: exactonline/http_test.py [hunk 10]
```diff
@@ -51,47 +58,65 @@ def __init__(self, method, code, body=None, use_ssl=False):
...
+            if response.body is None:
                 # If body is None, pass the indata as outdata.
                 body = data
                 if str != bytes:
```
Comment (AhsanPioneer) [inline exactonline/http_test.py:87]: Do we support both python3 and python2?

File: exactonline/http_test.py [hunk 11]
```diff
@@ -51,47 +58,65 @@ def __init__(self, method, code, body=None, use_ssl=False):
...
 
             peersock.send(
-                ('HTTP/1.0 %s Unused Response Title\r\n'
-                 'Content-Type: text/plain; utf-8\r\n'
-                 '\r\n%s' % (self.code, body)
-                 ).encode('utf-8'))
+                 ('HTTP/1.0 %s Unused Response Title\r\n'
+                  'Content-Type: text/plain; utf-8\r\n'
+                  '\r\n%s' % (response.code, body)
+                  ).encode('utf-8'))
```
Comment (AhsanPioneer) [inline exactonline/http_test.py:96]: Can use an f string here

File: exactonline/http_test.py [hunk 12]
```diff
@@ -51,47 +58,65 @@ def __init__(self, method, code, body=None, use_ssl=False):
...
             peersock.send(
                 ('HTTP/1.0 405 Method Not Implemented\r\n'
                  'Content-Type: text/plain; utf-8\r\n'
                  '\r\nUnexpected stuff'
                  ).encode('utf-8'))
+            error = True
+
         peersock.shutdown(self.SHUT_RDWR)
         peersock.close()
-        self.socket.close()  # server is done with it
```
Comment (AhsanPioneer) [inline exactonline/http_test.py:91]: Do we intend to close the socket somewhere else now?

File: exactonline/rawapi.py [hunk 13]
```diff
@@ -28,10 +31,109 @@ def _json_safe(data):
...
+
+    def backoff(self):
+        """
+        Check if we need to wait, and wait. Returns True if we did any waiting.
+        """
+        seconds = self._should_wait()
+        if seconds > 0:
+            self.wait(seconds)
+            return True
+        return False
```
Comment (AhsanPioneer) [inline exactonline/rawapi.py:60]:
```suggestion
            return True

        return False
```

File: exactonline/rawapi.py [hunk 14]
```diff
@@ -28,10 +31,109 @@ def _json_safe(data):
...
+            'Sleeping for %d seconds because of ratelimits %r',
+            seconds, self)
+
+        sleep(seconds)
+
+    def update(self, until, limit, remaining):
+        until = int(until)
+        limit = int(limit)
+        remaining = int(remaining)
+        assert 1638448971000 < until < 9999999999999, until
```
Comment (AhsanPioneer) [inline exactonline/rawapi.py:83]: Needs a comment here

File: exactonline/rawapi.py [hunk 15]
```diff
@@ -28,10 +31,109 @@ def _json_safe(data):
...
+        wait_until = None
+        amount_left = None
+        for key in self._reset_times:
+            if now < key:
+                if wait_until is None:
+                    wait_until = key
+                    amount_left = self._reset_times[key][1]
+                elif amount_left > self._reset_times[key][1]:
+                    wait_until = key
+                    amount_left = self._reset_times[key][1]
```
Comment (AhsanPioneer) [inline exactonline/rawapi.py:111]:
```suggestion
            if now > key:
                continue
            if wait_until is None:
                wait_until = key
                amount_left = self._reset_times[key][1]
            elif amount_left > self._reset_times[key][1]:
                wait_until = key
                amount_left = self._reset_times[key][1]
```

File: exactonline/rawapi.py [hunk 16]
```diff
@@ -146,23 +258,22 @@ def _rest_query(self, request):
...
+            response = http_req(
+                request.method, request.resource, data=request.data,
+                opt=opt, limiter=self.limiter)
+        except HTTPError as e:
+            if e.getcode() == 429 and self.limiter.backoff():
+                response = http_req(
+                    request.method, request.resource, data=request.data,
+                    opt=opt, limiter=self.limiter)
+            else:
+                raise
```
Comment (AhsanPioneer) [inline exactonline/rawapi.py:271]:
```suggestion
            raise
```

File: exactonline/rawapi.py [hunk 17]
```diff
@@ -171,11 +282,11 @@ def _set_tokens(self, jsondata):
         decoded = json.loads(jsondata)
 
         # Validate the values.
-        assert decoded['access_token']
-        expires_in = int(decoded['expires_in'])
-        assert expires_in > 0
-        assert decoded['refresh_token']
-        assert decoded['token_type'] == 'bearer'
+        assert decoded.get('access_token'), decoded
```
Comment (AhsanPioneer) [inline exactonline/rawapi.py:285]: Do we want to avoid raising key errors?

**Review by AhsanPioneer — COMMENTED** [review]
Nice work!
Just a few things which you can look into

**Comment by lrozema** [comment]
@AhsanPioneer maybe it should have been stated more clearly in the comments that this is simply merging in upstream changes. I don't think its worth the effort of improving that unless it actually breaks our stuff. It might actually be annoying to pull in future upstream change in terms of maitaining these improvements. Unless you see breaking things, shall we just merge this without applying the suggestions? Alternatively, we could open a PR towards Exact's upstream and hope they apply them, but again maybe not worth our efforts? 🤔

**Comment by AhsanPioneer** [comment]
@lrozema got it then we can go ahead and merge this one 
The comments are just for improvement

**Review by AhsanPioneer — APPROVED** [review]

**Pull request merged**